src/ring_theory/unique_factorization_domain.lean

```diff
@@ -1,18 +1,150 @@
+          apply lt_add_of_pos_right, apply nat.pos_of_ne_zero, contrapose! h, rw h, refl, }, }
+      end }
+  end }
+end prio
+
+namespace DCC_dvd
+
+variables [comm_monoid_with_zero α]
+open associates nat
+
+theorem of_DCC_dvd_associates
+  [DCC_dvd (associates α)]: DCC_dvd α :=
+⟨begin
+  refine (surjective.well_founded_iff mk_surjective _).2 DCC_dvd.well_founded_dvd_not_unit,
+  intros, rw mk_dvd_not_unit_mk_iff
+end⟩
+
+variables [DCC_dvd α]
+
+@[priority 100]
+instance DCC_dvd_associates : DCC_dvd (associates α) :=
+⟨begin
+  refine (surjective.well_founded_iff mk_surjective _).1 DCC_dvd.well_founded_dvd_not_unit,
+  intros, rw mk_dvd_not_unit_mk_iff
+end⟩
+
+theorem well_founded_associates : well_founded ((<) : associates α → associates α → Prop) :=
+subrelation.wf (λ x y, dvd_not_unit_of_lt) DCC_dvd.well_founded_dvd_not_unit
+
+local attribute [elab_as_eliminator] well_founded.fix
+
+lemma exists_irreducible_factor {a : α} (ha : ¬ is_unit a) (ha0 : a ≠ 0) :
+  ∃ i, irreducible i ∧ i ∣ a :=
+(irreducible_or_factor a ha).elim (λ hai, ⟨a, hai, dvd_refl _⟩)
+  (well_founded.fix
+    DCC_dvd.well_founded_dvd_not_unit
+    (λ a ih ha ha0 ⟨x, y, hx, hy, hxy⟩,
+      have hx0 : x ≠ 0, from λ hx0, ha0 (by rw [← hxy, hx0, zero_mul]),
+      (irreducible_or_factor x hx).elim
+        (λ hxi, ⟨x, hxi, hxy ▸ by simp⟩)
+        (λ hxf, let ⟨i, hi⟩ := ih x ⟨hx0, y, hy, hxy.symm⟩ hx hx0 hxf in
+          ⟨i, hi.1, dvd.trans hi.2 (hxy ▸ by simp)⟩)) a ha ha0)
+
+@[elab_as_eliminator] lemma induction_on_irreducible {P : α → Prop} (a : α)
+  (h0 : P 0) (hu : ∀ u : α, is_unit u → P u)
+  (hi : ∀ a i : α, a ≠ 0 → irreducible i → P a → P (i * a)) :
+  P a :=
+by haveI := classical.dec; exact
+well_founded.fix DCC_dvd.well_founded_dvd_not_unit
+  (λ a ih, if ha0 : a = 0 then ha0.symm ▸ h0
+    else if hau : is_unit a then hu a hau
+    else let ⟨i, hii, ⟨b, hb⟩⟩ := exists_irreducible_factor hau ha0 in
+      have hb0 : b ≠ 0, from λ hb0, by simp * at *,
+      hb.symm ▸ hi _ _ hb0 hii (ih _ ⟨hb0, i,
+        hii.1, by rw [hb, mul_comm]⟩))
+  a
+
+lemma exists_factors (a : α) : a ≠ 0 →
+  ∃f : multiset α, (∀b ∈ f, irreducible b) ∧ associated a f.prod :=
+DCC_dvd.induction_on_irreducible a
+  (λ h, (h rfl).elim)
+  (λ u hu _, ⟨0, by simp [associated_one_iff_is_unit, hu]⟩)
+  (λ a i ha0 hii ih hia0,
+    let ⟨s, hs⟩ := ih ha0 in
+    ⟨i::s, ⟨by clear _let_match; finish,
+      by rw multiset.prod_cons;
+        exact associated_mul_mul (by refl) hs.2⟩⟩)
+
+end DCC_dvd
+
+theorem DCC_dvd_of_well_founded_associates [comm_cancel_monoid_with_zero α]
+  (h : well_founded ((<) : associates α → associates α → Prop)): DCC_dvd α :=
+@DCC_dvd.of_DCC_dvd_associates _ _ ⟨by { convert h, ext, exact associates.dvd_not_unit_iff_lt }⟩
+
+theorem DCC_dvd_iff_well_founded_associates [comm_cancel_monoid_with_zero α] :
+  DCC_dvd α ↔ well_founded ((<) : associates α → associates α → Prop) :=
+⟨by apply DCC_dvd.well_founded_associates, DCC_dvd_of_well_founded_associates⟩
+
```

Comment:
Would it be hard to add a theorem showing the equivalence between `DCC_dvd` and `unique_factorization_domain` (or at least one of the implications), so any new code can make use of `DCC_dvd` without having to port the lemmas from `unique_factorization_domain`?

Reply:
Sorry, that should be "equivalence between the future definition of `unique_factorization_domain` that uses `DCC_dvd` and the current definition".

Reply:
I have coded such a theorem, but it uses my new `unique_factorization_monoid` class, and I feel like adding that would either double the length of the PR, or do it in a "wrong" way that I'd want to completely replace with my next PR anyway.

Reply:
The rest of my work is on the branch `mathlib:ufm` if you're interested.

Reply:
Doubling the PR size or writing a lot of code to be thrown away soon is definitely too much work, so you were right to not include it in the first place :)